File: matlab/ReadRSS.m
function [data, head] = ReadRSS(filename)
% ReadRSS : Reads a RSS formatted file
%
% Call :
% [Data,Header]=ReadRSS(filename);
% Example:
% [Data, Header] = ReadRSS('../Test/MPI/Acoustic2D/Pshot.rss');
% ot = Header.geometry.O(1); % Time origin
% dt = Header.geometry.D(1); % Sampling interval 
% nt = Header.geometry.N(1); % Number of time samples
% ntr = Header.geometry.N(2); % Number of traces
% SrcCoords = [Header.coordinates(1:ntr).s]; % Source coordinates (x,y,z)
% RecCoords = [Header.coordinates(1:ntr).g]; % Receiver coordinates (x,y,z)
% % Display data:
% t = (0:nt-1)*dt + ot; 
% x = RecCoords(1:Header.Nheader/2:end);
% figure, imagesc(1:ntr, t, Data); 
% set(gca, 'xticklabel', x(get(gca,'xtick')));
%
%   Author(s): W. Weibull, 10/09-17
%   Copyright 2017 Wiktor Weibull

MAGICNUMBER='R0CKS'; %rss identifier
MAGICNUMBERLENGTH=5;
MAXDIMS=9;
data=[];
head=[];

fileID = fopen(filename, 'r');
% Check if file has rss identifier
buff = fread(fileID,MAGICNUMBERLENGTH,'*char')';
if(strcmp(buff, MAGICNUMBER) == false)
    display('Not a valid rss file!');
    return;
end

% Reading header
head.data_format = fread(fileID, 1,'*int');
head.header_format = fread(fileID, 1,'*int');
head.type = fread(fileID, 1,'*int');
head.Nheader = fread(fileID, 1,'*uint64');
head.Ndims = fread(fileID, 1,'*uint64');

for i=1:MAXDIMS
    head.geometry.N(i) = fread(fileID, 1,'*uint64');
end

for i=1:MAXDIMS
    head.geometry.D(i) = fread(fileID, 1,'*float64');
end

for i=1:MAXDIMS
    head.geometry.O(i) = fread(fileID, 1,'*float64');
end

% Reading traces and coordinates
if(head.Nheader)
    data=zeros(head.geometry.N(1), head.geometry.N(2));
    for i=1:head.geometry.N(2)
        for j=1:head.Nheader/2
            head.coordinates(i).s(j) = fread(fileID, 1,'*float32');
        end
        for j=1:head.Nheader/2
            head.coordinates(i).g(j) = fread(fileID, 1,'*float32');
        end
        
        data(:,i) = fread(fileID, head.geometry.N(1),'*float32');
    end
else
    for i=1:MAXDIMS
        if (head.geometry.N(i) > 0)
            Dsz(i) = head.geometry.N(i);
        end
    end
    data = fread(fileID, prod(Dsz),'*float32');
    data = reshape(data, Dsz);
end

fclose(fileID);

end

File: python/ReadRSS.py
import numpy as np
import struct as st

MAGICNUMBER='R0CKS' #rss identifier
MAGICNUMBERLENGTH=5
MAXDIMS=9

class RSSdata:
    def __init__(self):
        self.data_format=0;
        self.header_format=0;
        self.type=0;
        self.Nheader=0;
        self.Ndims=0;
        self.geomN = np.zeros([MAXDIMS,1], dtype='uint64')
        self.geomD = np.zeros([MAXDIMS,1], dtype='float64')
        self.geomO = np.zeros([MAXDIMS,1], dtype='float64')

    def read(self,filename):
        f = open(filename, 'rb')
        buff= f.read(MAGICNUMBERLENGTH)
    
        # Check if file has rss identifier
        if(buff != MAGICNUMBER):
            print 'Not a valid rss file!'
            return -1
    
        # Reading header
        head = {}
        self.data_format = st.unpack('i', f.read(4))[0]
        self.header_format = st.unpack('i', f.read(4))[0]
        self.type = st.unpack('i', f.read(4))[0]
        self.Nheader = st.unpack('Q', f.read(8))[0]
        self.Ndims = st.unpack('Q', f.read(8))[0]
        for i in range(0,MAXDIMS):
            self.geomN[i] = st.unpack('Q', f.read(8))[0]
        for i in range(0,MAXDIMS):
            self.geomD[i] = st.unpack('d', f.read(8))[0]
        for i in range(0,MAXDIMS):
            self.geomO[i] = st.unpack('d', f.read(8))[0]

        #Reading coordinates and data
        fullsize = 1;
        self.Ndims = 0
        for i in range(0, MAXDIMS):
            if(self.geomN[i] > 0):    
                fullsize = fullsize*self.geomN[i];
                self.Ndims = self.Ndims + 1
        resize = np.zeros([self.Ndims,1], dtype='uint64')
        for i in range(0,self.Ndims):
            resize[i] = self.geomN[i];

        # Allocate 1d array for data
        self.data = np.zeros([fullsize,1], dtype='float32')
        if(self.Nheader):
            self.srcX = np.zeros([self.geomN[1],1], dtype='float32')
            self.srcY = np.zeros([self.geomN[1],1], dtype='float32')
            self.srcZ = np.zeros([self.geomN[1],1], dtype='float32')
            self.GroupX = np.zeros([self.geomN[1],1], dtype='float32')
            self.GroupY = np.zeros([self.geomN[1],1], dtype='float32')
            self.GroupZ = np.zeros([self.geomN[1],1], dtype='float32')
            for i in range(0,self.geomN[1]):
                if(self.Nheader == 4):
                    self.srcX[i] = st.unpack('f', f.read(4))[0];
                    self.srcZ[i] = st.unpack('f', f.read(4))[0];
                    self.GroupX[i] = st.unpack('f', f.read(4))[0];
                    self.GroupZ[i] = st.unpack('f', f.read(4))[0];
        
                if(self.Nheader == 6):
                    self.srcX[i] = st.unpack('f', f.read(4))[0];
                    self.srcY[i] = st.unpack('f', f.read(4))[0];
                    self.srcZ[i] = st.unpack('f', f.read(4))[0];
                    self.GroupX[i] = st.unpack('f', f.read(4))[0];
                    self.GroupY[i] = st.unpack('f', f.read(4))[0];
                    self.GroupZ[i] = st.unpack('f', f.read(4))[0];
                            
                for j in range(0,self.geomN[0]):
                    self.data[i*self.geomN[0] + j] = st.unpack('f', f.read(4))[0];
        else:
            for i in range(1,fullsize):
                self.data[i] = st.unpack('f', f.read(4))[0];

        # Reshape data to correct dimensions
        self.data=self.data.reshape(resize)
        print self.data.shape

